File: vercors/src/main/java/vct/col/ast/ParallelBarrier.java
package vct.col.ast;

import java.util.ArrayList;

public class ParallelBarrier extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 
  public final String label;
  
  public final Contract contract;
  
  public final ArrayList<String> invs;
  
  public final BlockStatement body;
  
  public ParallelBarrier(String label,Contract contract, ArrayList<String> fences, BlockStatement body){
    this.label=label;
    this.contract=contract;
    this.invs=new ArrayList<String>(fences);
    this.body=body;
  }

}

File: vercors/src/main/java/vct/col/ast/ParallelBarrier.scala
package vct.col.ast

import java.util.ArrayList
import scala.collection.JavaConverters._
import vct.col.util.VisitorHelper

class ParallelBarrier (val label:String, val contract:Contract, private[this] val fences:ArrayList[String], val body:BlockStatement) extends ASTNode with VisitorHelper {
  val invs = new ArrayList[String](fences)
    
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
}
